File: src/billing_mixin/NightlyDiscountPolicy.scala
package billing_mixin

import java.time.Duration

import money.Money

class NightlyDiscountPolicy(
                             val nightlyAmount: Money,
                             val regularAmount: Money,
                             val seconds: Duration) extends BasicRatePolicy {

  override protected def calculateCallFee(call: Call): Money =
    if (call.from.getHour >= NightltDiscountPolicy.LateNightHour) {
      nightlyAmount.times(call.duration.getSeconds / seconds.getSeconds)
    } else {
      regularAmount.times(call.duration.getSeconds / seconds.getSeconds)
    }
}

object NightltDiscountPolicy {
  val LateNightHour: Integer = 22
}

File: src/billing_composition/step01/NightlyDiscountPolicy.java
package billing_composition.step01;

import money.Money;

import java.time.Duration;

public class NightlyDiscountPolicy extends BasicRatePolicy {
    private static final int LATE_NIGHT_HOUR = 22;

    private Money nightlyAmount;
    private Money regularAmount;
    private Duration seconds;

    public NightlyDiscountPolicy(Money nightlyAmount, Money regularAmount, Duration seconds) {
        this.nightlyAmount = nightlyAmount;
        this.regularAmount = regularAmount;
        this.seconds = seconds;
    }

    @Override
    protected Money calculateCallFee(Call call) {
        long percent = call.getDuration().getSeconds() / seconds.getSeconds();
        if (call.getFrom().getHour() >= LATE_NIGHT_HOUR) {
            return nightlyAmount.times(percent);
        }
        return regularAmount.times(percent);
    }
}
